### PR TITLE
Themes: Incorrect template part slugs when the theme folder name ends with `parts`

### PR DESCRIPTION
**Pull request by yashjawale**
Trac ticket: https://core.trac.wordpress.org/ticket/65580

## Use of AI Tools

AI assistance: Yes
Tool(s): OpenCode
Model(s): DeepSeek V4 Flash
Used for: Checking initial implementation, followed by manual testing & verification

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

**Review by t-hamano — COMMENTED** [review]
Thanks for the PR. Could you add unit tests to ensure all scenarios are covered and no regressions occur? It would be good to cover as many edge cases as possible.

**Comment by yashjawale** [comment]
Added unit tests for template/template part slugs with `parts` in folder names ✅ 

Also with the tests noticed that similar problem arised for template parts in nested folders, so I've updated the approach from `strrpos` to stripping out the prefix of theme directory + template base path from the file path with length instead to get the slug.

File: tests/phpunit/data/themedir1/block-theme-ending-with-parts/parts/footer.html
Comment (t-hamano) [inline tests/phpunit/data/themedir1/block-theme-ending-with-parts/parts/footer.html:1]: Personally, I'm hesitant to add so many HTML files just for unit testing. The `_get_block_templates_files()` function retrieves the theme directory via `get_stylesheet_directory()`, so it might be replaced with a temporary directory using the `stylesheet_directory` filter. It's probably something like this.

```php
/**
 * @dataProvider data_get_block_templates_files_slugs
 */
public function test_get_block_templates_files_slugs( $theme_directory, $template_type, $expected_slugs ) {
	add_filter(
		'stylesheet_directory',
		static function () use ( $theme_directory ) {
			return self::$slug_theme_root . '/' . $theme_directory;
		}
	);

	$slugs = wp_list_pluck( _get_block_templates_files( $template_type ), 'slug' );
	sort( $slugs );

	$this->assertSame( $expected_slugs, $slugs );
}

public function data_get_block_templates_files_slugs() {
	return array(
		'template parts in a theme directory named after the template part base path' => array(
			'theme_directory' => 'parts',
			'template_type'   => 'wp_template_part',
			'expected_slugs'  => array( 'footer-parts/header', 'header' ),
		),
		// ...
	);
}
```

Comment (yashjawale) [reply tests/phpunit/data/themedir1/block-theme-ending-with-parts/parts/footer.html:1]: Thanks a lot! 
I've updated the PR to use temp directories & its much cleaner.
Not sure why some tests are failing at exactly 20min mark tho... will retrigger them later.

File: tests/phpunit/tests/block-template.php [hunk 1]
```diff
@@ -22,6 +64,8 @@ public function tear_down() {
 		global $_wp_current_template_id, $_wp_current_template_content;
 		unset( $_wp_current_template_id, $_wp_current_template_content );
 
+		remove_all_filters( 'stylesheet_directory' );
+
```
Comment (t-hamano) [inline tests/phpunit/tests/block-template.php]:
```suggestion

```

Nit: Since `WP_UnitTestCase_Base::tear_down()` calls `_restore_hooks()`, this should be unnecessary.

https://github.com/t-hamano/wordpress-develop/blob/33492385eead9d49756a3e7d2d0870e30962cdd0/tests/phpunit/includes/abstract-testcase.php#L230

File: tests/phpunit/tests/block-template.php [hunk 2]
```diff
@@ -11,6 +11,48 @@ class Tests_Block_Template extends WP_UnitTestCase {
...
+	public static function wpSetUpBeforeClass( WP_UnitTest_Factory $factory ) {
+		self::$slug_theme_root = sys_get_temp_dir() . '/wp-tests-block-template-slug-' . uniqid();
+
+		$files = array(
+			'parts/header.html'              => '<!-- wp:template-part /-->',
+			'parts/footer.html'              => '<!-- wp:template-part /-->',
+			'parts/footer-parts/header.html' => '<!-- wp:template-part /-->',
+			'templates/index.html'           => '<!-- wp:template-part /-->',
+			'templates/footer.html'          => '<!-- wp:template-part /-->',
+		);
```
Comment (t-hamano) [inline tests/phpunit/tests/block-template.php]: The contents of these array elements are actually impossible and likely incorrect. Furthermore, these contents are not used in testing.

This should be simplified as follows.

```diff
diff --git a/tests/phpunit/tests/block-template.php b/tests/phpunit/tests/block-template.php
index f164b6d5e6..c4c8c9a1e2 100644
--- a/tests/phpunit/tests/block-template.php
+++ b/tests/phpunit/tests/block-template.php
@@ -21,21 +21,21 @@ class Tests_Block_Template extends WP_UnitTestCase {
        public static function wpSetUpBeforeClass( WP_UnitTest_Factory $factory ) {
                self::$slug_theme_root = sys_get_temp_dir() . '/wp-tests-block-template-slug-' . uniqid();
 
-               $files = array(
-                       'parts/header.html'              => '<!-- wp:template-part /-->',
-                       'parts/footer.html'              => '<!-- wp:template-part /-->',
-                       'parts/footer-parts/header.html' => '<!-- wp:template-part /-->',
-                       'templates/index.html'           => '<!-- wp:template-part /-->',
-                       'templates/footer.html'          => '<!-- wp:template-part /-->',
+               $relative_paths = array(
+                       'parts/header.html',
+                       'parts/footer.html',
+                       'parts/footer-parts/header.html',
+                       'templates/index.html',
+                       'templates/footer.html',
                );
 
                // The same template tree is created under each directory name used by
                // the data provider, so that only the theme directory name varies.
                foreach ( array( 'default', 'auto-parts', 'parts', 'my-templates' ) as $theme_directory ) {
-                       foreach ( $files as $relative_path => $content ) {
+                       foreach ( $relative_paths as $relative_path ) {
                                $file_path = self::$slug_theme_root . '/' . $theme_directory . '/' . $relative_path;
                                wp_mkdir_p( dirname( $file_path ) );
-                               file_put_contents( $file_path, $content );
+                               touch( $file_path );
                        }
                }
        }
```

File: tests/phpunit/tests/block-template.php [hunk 3]
```diff
@@ -11,6 +11,48 @@ class Tests_Block_Template extends WP_UnitTestCase {
...
+			if ( $item->isDir() ) {
+				rmdir( $item->getPathname() );
+			} else {
+				unlink( $item->getPathname() );
+			}
+		}
+		rmdir( self::$slug_theme_root );
+
+		self::$slug_theme_root = null;
+	}
```
Comment (t-hamano) [inline tests/phpunit/tests/block-template.php]: This approach to deleting temporary directories is not present in other unit tests. Perhaps the approach that is most frequently used, as shown below, would be better.

```php
public function test_get_block_templates_files_slugs( $theme_directory, $template_type, $expected_slugs ) {
	$theme_root = sys_get_temp_dir() . '/wp-tests-block-template-slug-' . uniqid();
	$theme_dir  = $theme_root . '/' . $theme_directory;

	foreach ( $files as $file_path ) {
		wp_mkdir_p( dirname( $file_path ) );
		touch( $file_path );
	}

	add_filter(
		'stylesheet_directory',
		static function () use ( $theme_dir ) {
			return $theme_dir;
		}
	);

	$slugs = wp_list_pluck( _get_block_templates_files( $template_type ), 'slug' );
	sort( $slugs );

	foreach ( $files as $file_path ) {
		unlink( $file_path );
	}
	foreach ( $directories as $directory ) {
		rmdir( $directory );
	}

	$this->assertSame( $expected_slugs, $slugs );
}
```

File: tests/phpunit/tests/block-template.php [hunk 4]
```diff
@@ -509,6 +553,77 @@ public function test_get_block_templates_paths_dir_doesnt_exists() {
...
+			'template parts in a theme whose directory name is exactly parts' => array(
+				'theme_directory' => 'parts',
+				'template_type'   => 'wp_template_part',
+				'expected_slugs'  => array( 'footer', 'footer-parts/header', 'header' ),
+			),
+			'templates in a theme whose directory name ends with templates' => array(
+				'theme_directory' => 'my-templates',
+				'template_type'   => 'wp_template',
+				'expected_slugs'  => array( 'footer', 'index' ),
+			),
```
Comment (t-hamano) [inline tests/phpunit/tests/block-template.php:617]:
```suggestion
			),
			'template parts in a theme placed inside a directory named parts' => array(
				'theme_directory' => 'parts/default',
				'template_type'   => 'wp_template_part',
				'expected_slugs'  => array( 'footer', 'footer-parts/header', 'header' ),
			),
			'templates in a theme placed inside a directory named templates' => array(
				'theme_directory' => 'templates/default',
				'template_type'   => 'wp_template',
				'expected_slugs'  => array( 'footer', 'index' ),
			),
```

It might be better to also test cases where the directory above the theme includes `parts` and `templates`.

File: tests/phpunit/tests/block-template.php (hunk 2 above)
Comment (yashjawale) [reply tests/phpunit/tests/block-template.php]: Ah... sorry for missing that.
In the new approach I've updated it to create empty files instead of with content.

File: tests/phpunit/tests/block-template.php (hunk 3 above)
Comment (yashjawale) [reply tests/phpunit/tests/block-template.php]: Thanks! It looks a lot better.
Updated with the change.

**Comment by yashjawale** [comment]
Updated the PR but looks like CI checks are facing some issues...
I'll retrigger them after a while.

**Review by t-hamano — APPROVED** [review]
LGTM!
